### ISSUE TITLE
Sharing Population and Switching Algorithms.

### ISSUE THREAD

**Issue by vsantjr** [jMetal/jMetal#246]
Hello,

Let us suppose that I have two metaheuristics (Alg1, Alg2) and I have to share the population between them and, in addition, to switch the execution of both algorithms per generation of a new population. In order words, the flow of execution is something like this:

Pop0 (initial random) -> Alg1 -> Pop1 (new population) -> Alg 2 -> Pop2 (new population) -> Alg1 -> Pop3 (new population) ...

In other words, Alg1 generates a population (Pop1), after that Alg2 generates another one (Pop2) based on the previous population Pop1 generated by Alg1, and so on.

Any ideas on how to do this with jMetal? 

Thanks again.

**Comment by matthieu-vergne** [comment]
You can always try to hack the algorithms, but basically they are implemented to run at once, not to be paused, updated, and restarted. So they are not naturally suited to this kind of interactions. However, if the measures are implemented, you can try to build on that.

The basic idea would be to setup the first algorithm to be notified for each new generation. When you are notified, you get the new population, give it to the second algorithm, run it for 1 generation, get the result, and put it back into the first algorithm. Then the first algorithm continues, you get the next notification, which redo that again, and so on. Such a process would need:
- the first algorithm to notifies of new populations with the measures,
- the second algorithm to fulfil the property : 1 run of N generations = N runs of 1 generation

If your algorithms fulfil these conditions, you may use that strategy.

If they don't, but both implement measures, you may try something trickier by getting notified of new populations on both, and to implement some kind of synchronisation mechanism to ensure that each algorithm wait for you to make another run with the other before to continue. It is not easy, but should be doable, although it requires som tricky design.

Finally, if one of the algorithms you use is setup to run in parallel, then just forget about it. All which remains to you is probably to implement your own algorithm.

**Comment by ajnebro** [comment]
You could run each algorithm in its own thread and use the producer-consumer model to synchronize the use of the population alternatively by each of them.

**Comment by vsantjr** [comment]
Ok, Thanks. I will try some of these ideas and get back if I have additional doubts.

**Comment by vsantjr** [comment]
Regarding this point, I compared NSGAIIMeasures x DMOPSOMeasures. In NSGAII, we have this code related to hypervolume (initMeasures):

measureManager.setPushMeasure("hypervolume", hypervolumeValue);

But in DMPOSO we have both pull and push (initMeasures):

measureManager.setPullMeasure("hypervolume", hypervolumeValue);

measureManager.setPushMeasure("hypervolume", hypervolumeValue);

Thus do I need to pull and push each quality indicator within the initMeasures or only push? I just need to calculate the quality indicator for each generation or some set of generations.

Thanks.

**Comment by matthieu-vergne** [comment]
Measures are here to give access to stuff. Push measures are active (they generate notifications), while pull measures are passive (they are mere getters). If both are available, just use the one you prefer.

**Comment by vsantjr** [comment]
Thank you.
